### PR TITLE
feat(cycle-details-endpoint)

### PR DESCRIPTION
**Pull request by iaheck**
# Base PR
- PR based on main

# Context
- With this PR, we are adding the endpoint `api/cycles/:id` to retrieve the details of a cycle: 

image

This is part of our first slice of work.

# Changelog
- Create a new route for the endpoint, and a controller to manage it
- Create tests for the endpoint and swagger documentation
- It also includes a new API error handler for `ActiveRecord::RecordInvalid`, returning an error 400

# QA
- Make sure of have run the seeds and being logged in
- Go to `http://localhost:3000/api/cycles/1` to see the endpoint output

### CODE REVIEW

File: spec/requests/api/cycles_spec.rb [hunk 1]
```diff
@@ -42,4 +42,42 @@
       end
     end
   end
+
+  path '/api/cycles/{id}' do
+    get 'Returns Cycle details' do
+      tags 'Cycles'
+      description 'Retrieves the details of a cycle'
+      parameter name: :curriculum_id, in: :path, type: :string
```
Comment (AlehandroL) [inline spec/requests/api/cycles_spec.rb:50]: this line should be deleted entirely

File: spec/requests/api/cycles_spec.rb [hunk 2]
```diff
@@ -42,4 +42,42 @@
...
+                 order_number: { type: :integer },
+                 learning_goals_description: { type: :string },
+                 challenge_description: { type: :string },
+                 boilerplate_url: { type: :string }
+               }
+
+        run_test!
+      end
+
+      response '401', 'Unauthorized', skip_before: true do
```
Comment (AlehandroL) [inline spec/requests/api/cycles_spec.rb:73]: a cycle information should be public to users that are not authenticated, following [this PR's guidelines](https://github.com/fintual-oss/paraffin-back/pull/10), therefore you should delete this test

File: swagger/swagger.json [hunk 3]
```diff
@@ -135,6 +135,72 @@
...
+                    "boilerplate_url": {
+                      "type": "string"
+                    }
+                  }
+                }
+              }
+            }
+          },
+          "401": {
+            "description": "Unauthorized"
```
Comment (AlehandroL) [inline swagger/swagger.json:196]: please update after deleting this test

File: app/controllers/api/api_application_controller.rb [hunk 4]
```diff
@@ -21,6 +22,19 @@ def record_not_found
     status = :not_found
     code = 404
     message = 'Record_not_found'
-    render json: { status:, code:, message: }, status:
+
+    render_error(code, message, status)
+  end
+
+  def handle_record_invalid_error
+    code = 400
```
Comment (AlehandroL) [inline app/controllers/api/api_application_controller.rb:30]: I understand the need to handle this error, but since this method isn't used in this PR, I think it should be on a different PR, along with the places it is called

**Review by AlehandroL — CHANGES_REQUESTED** [review]
Looks great, only a few minor changes.

File: spec/requests/api/cycles_spec.rb (hunk 1 above)
Comment (iaheck) [reply spec/requests/api/cycles_spec.rb:50]: you're right, thanks!

File: spec/requests/api/cycles_spec.rb (hunk 2 above)
Comment (iaheck) [reply spec/requests/api/cycles_spec.rb:69]: done! thanks

File: swagger/swagger.json (hunk 3 above)
Comment (iaheck) [reply swagger/swagger.json:187]: done

File: app/controllers/api/api_application_controller.rb (hunk 4 above)
Comment (iaheck) [reply app/controllers/api/api_application_controller.rb]: alright, I removed that commit for now

**Review by AlehandroL — APPROVED** [review]

File: app/controllers/api/cycles_controller.rb [hunk 5]
```diff
@@ -1,17 +1,26 @@
...
+    before_action :set_cycle, only: [:show]
 
     def index
       cycles = @curriculum.cycles
       render json: cycles,
              only: %i[id name learning_goals_description]
     end
 
+    def show
+      render json: @cycle, except: %i[created_at updated_at curriculum_id]
```
Comment (PabloAbara) [inline app/controllers/api/cycles_controller.rb:13]: Suggestion: Could be better to explicitly declare what includes instead of what excludes.

File: spec/requests/api/cycles_spec.rb [hunk 6]
```diff
@@ -38,4 +38,37 @@
...
+                 order_number: { type: :integer },
+                 learning_goals_description: { type: :string },
+                 challenge_description: { type: :string },
+                 boilerplate_url: { type: :string }
+               }
+
+        run_test!
+      end
+
+      response '404', 'Cycle not found' do
```
Comment (PabloAbara) [inline spec/requests/api/cycles_spec.rb:68]: I'm not sure if you should test Unauthorized response. It depends if this access is public or not.

**Review by PabloAbara — APPROVED** [review]

File: app/controllers/api/cycles_controller.rb (hunk 5 above)
Comment (iaheck) [reply app/controllers/api/cycles_controller.rb]: done, thanks!

File: spec/requests/api/cycles_spec.rb (hunk 6 above)
Comment (iaheck) [reply spec/requests/api/cycles_spec.rb:68]: you won't need to be logged in to access this

**Review by PabloAbara — APPROVED** [review]

**Review by AlehandroL — APPROVED** [review]
